pattern.c

```diff
@@ -1334,17 +1335,16 @@ static /* const */ char *find_matching_paren(/* const */ char *s)
  * mutt_pattern_free - Free a Pattern
  * @param pat Pattern to free
  */
-void mutt_pattern_free(struct Pattern **pat)
+void mutt_pattern_free(struct PatternHead **pat)
 {
   if (!pat || !*pat)
     return;
 
   struct Pattern *tmp = NULL;
-
-  while (*pat)
+  while (!SLIST_EMPTY(*pat))
```

Comment:
Here is the code of both:

SLIST_EMPTY:

```c
void mutt_pattern_free(struct PatternHead **pat)
{
  if (!pat || !*pat)
    return;

  struct Pattern *tmp = NULL;
  while (!SLIST_EMPTY(*pat))
  {
    tmp = SLIST_FIRST(*pat);
    SLIST_REMOVE_HEAD(*pat, entries);

    if (tmp->ismulti)
      mutt_list_free(&tmp->p.multi_cases);
    else if (tmp->stringmatch)
      FREE(&tmp->p.str);
    else if (tmp->groupmatch)
      tmp->p.group = NULL;
    else if (tmp->p.regex)
    {
      regfree(tmp->p.regex);
      FREE(&tmp->p.regex);
    }

    mutt_pattern_free(&tmp->child);
    FREE(&tmp);
  }


  FREE(pat);
}
```

SLIST_FOREACH_SAFE:

```c
/**
 * mutt_pattern_free - Free a Pattern
 * @param[out] pat Pattern to free
 */
void mutt_pattern_free(struct PatternHead **pat)
{
  if (!pat || !*pat)
    return;

  struct Pattern *tmp1 = NULL, *tmp2 = NULL;
  SLIST_FOREACH_SAFE(tmp1, *pat, entries, tmp2) {
    SLIST_REMOVE(*pat, tmp1, Pattern, entries);

    if (tmp1->ismulti)
      mutt_list_free(&tmp1->p.multi_cases);
    else if (tmp1->stringmatch)
      FREE(&tmp1->p.str);
    else if (tmp1->groupmatch)
      tmp1->p.group = NULL;
    else if (tmp1->p.regex)
    {
      regfree(tmp1->p.regex);
      FREE(&tmp1->p.regex);
    }

    mutt_pattern_free(&tmp1->child);
    FREE(&tmp1);
  }

  FREE(pat);
}
```

The SLIST_FOREACH_SAFE performs an SLIST_REMOVE which is O(n) even though it will always remove the head so O(1). In the SLIST_EMPTY case, the removal is clearly O(1). I like the explicitness of SLIST_EMPTY, but I'll leave you with the decision.

Reply:
I would use the same pattern that's being used in `mutt_list_free`: https://github.com/neomutt/neomutt/blob/40abdccd1519655dfdf7ef36fd407df7d15368ed/mutt/list.c#L105

Reply:
Just pushed changes, should be identical to mutt_list_free.